### ISSUE TITLE
Provide official Docker image for murmur

### ISSUE THREAD

**Issue by gclawes** [mumble-voip/mumble#3688]
It would be nice to have an official image of murmur for containerized deployments.  I can take a crack at this if I get some guidance for what version to target (1.2.x or 1.3.x)

**Comment by davidebeatrici** [comment]
We should target 1.3.x, feel free to ask if you have any questions regarding the build process.

**Comment by gclawes** [comment]
Thanks.  A few things to clarify to start with:
* Any preference for base OS?  My inclination would be ubuntu xenial, based on what's in `.travis.yml`
* Qt4 or Qt5?
* Enable gRPC? Or build a separate grpc-enabled image?

**Comment by davidebeatrici** [comment]
You're welcome!

- Choose the distribution you prefer.
- Qt 5.
- Enable gRPC.
We didn't do it in our 1.3 builds because unfortunately quite some changes are required in our current build environment.

**Comment by gclawes** [comment]
I'll try a multistage build, compiling a static binary in fedora:30 and installing to alpine to keep the image size small.

**Comment by davidebeatrici** [comment]
Sounds good!

**Comment by gclawes** [comment]
ICE isn't available on `fedora` anymore, going with `ubuntu:disco`

**Comment by davidebeatrici** [comment]
No problem.

**Comment by gclawes** [comment]
It looks like for `static` builds, [src/murmur/murmur_ice/murmur_ice.pro](https://github.com/mumble-voip/mumble/blob/master/src/murmur/murmur_ice/murmur_ice.pro#L59) always expects slice2cpp to be at `/opt/Ice-3.3/bin/slice2cpp`, throwing an error when it isn't there.  Is there a reason for static builds hard-coding this?

```
Step 7/7 : RUN make release
 ---> Running in a159492cb0ff
cd src/mumble_proto/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/mumble_proto/mumble_proto.pro CONFIG+=no-client CONFIG+=grpc CONFIG+=static ) && make -f Makefile release
make[1]: Entering directory '/root/mumble/src/mumble_proto'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/mumble_proto'
/usr/bin/protoc --cpp_out=. -I. -I.. -I.. ../Mumble.proto
g++ -c -pipe -fvisibility=hidden -Wall -Wextra -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -O2 -fPIC -std=gnu++1z -w -pthread -DRESTRICT=__restrict__ -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../release/.obj/mumble_proto/Mumble.pb.o Mumble.pb.cc
rm -f ../../release/libmumble_proto.a
ar cqs ../../release/libmumble_proto.a ../../release/.obj/mumble_proto/Mumble.pb.o
make[2]: Leaving directory '/root/mumble/src/mumble_proto'
make[1]: Leaving directory '/root/mumble/src/mumble_proto'
cd src/murmur/murmur_ice/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur_ice/murmur_ice.pro CONFIG+=no-client CONFIG+=grpc CONFIG+=static ) && make -f Makefile release
make[1]: Entering directory '/root/mumble/src/murmur/murmur_ice'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/murmur/murmur_ice'
/opt/Ice-3.3/bin/slice2cpp --checksum -I/opt/Ice-3.3/slice ../Murmur.ice
make[2]: /opt/Ice-3.3/bin/slice2cpp: Command not found
make[2]: *** [Makefile.Release:188: Murmur.cpp] Error 127
make[2]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
make[1]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
make[1]: *** [Makefile:42: release] Error 2
make: *** [Makefile:451: sub-src-murmur-murmur_ice-release_ordered] Error 2
The command '/bin/sh -c make release' returned a non-zero code: 2
```
Ubuntu installs it at `/usr/bin/slice2cpp`

**Comment by gclawes** [comment]
This is configured here as well: https://github.com/mumble-voip/mumble/blob/master/src/murmur/murmur.pro#L139

**Comment by davidebeatrici** [comment]
Feel free to delete those lines, they were added more than a decade ago in commit f4064dfe8d076bab5c16d5b93983e0ec1c3a269a (a dedicated project for Ice was created later, in d8889f8bde96d1d998e5c25af6886aede5986889).

**Comment by gclawes** [comment]
Thanks, will do.

**Comment by gclawes** [comment]
I'm running into an issue building the GRPC support:
```Step 7/7 : RUN make
 ---> Running in bc9370b71c51
cd src/mumble_proto/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/mumble_proto/mumble_proto.pro 'CONFIG+=no-client grpc static' ) && make -f Makefile
make[1]: Entering directory '/root/mumble/src/mumble_proto'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/mumble_proto'
/usr/bin/protoc --cpp_out=. -I. -I.. -I.. ../Mumble.proto
g++ -c -pipe -fvisibility=hidden -Wall -Wextra -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -O2 -fPIC -std=gnu++1z -w -pthread -DRESTRICT=__restrict__ -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../release/.obj/mumble_proto/Mumble.pb.o Mumble.pb.cc
rm -f ../../release/libmumble_proto.a
ar cqs ../../release/libmumble_proto.a ../../release/.obj/mumble_proto/Mumble.pb.o
make[2]: Leaving directory '/root/mumble/src/mumble_proto'
make[1]: Leaving directory '/root/mumble/src/mumble_proto'
cd src/murmur/murmur_ice/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur_ice/murmur_ice.pro 'CONFIG+=no-client grpc static' ) && make -f Makefile
make[1]: Entering directory '/root/mumble/src/murmur/murmur_ice'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/murmur/murmur_ice'
slice2cpp --checksum -I/usr/local/share/Ice -I/usr/share/Ice/slice -I/usr/share/ice/slice -I/usr/share/slice -I/usr/share/Ice-3.7.2/slice/ ../Murmur.ice
g++ -c -pipe -fvisibility=hidden -Wall -Wextra -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -O2 -fPIC -std=gnu++1z -w -DRESTRICT=__restrict__ -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../../release/.obj/murmur_ice/Murmur.o Murmur.cpp
rm -f ../../../release/libmurmur_ice.a
ar cqs ../../../release/libmurmur_ice.a ../../../release/.obj/murmur_ice/Murmur.o
make[2]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
make[1]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
cd src/murmur_grpcwrapper_protoc_plugin/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur_grpcwrapper_protoc_plugin/murmur_grpcwrapper_protoc_plugin.pro 'CONFIG+=no-client grpc static' ) && make -f Makefile
make[1]: Entering directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
g++ -c -pipe -fvisibility=hidden -Wall -Wextra -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I../mumble_proto -isystem ../mumble_proto -O2 -std=gnu++1z -pthread -Wall -W -D_REENTRANT -fPIC -DRESTRICT=__restrict__ -DMUMBLE_VERSION_STRING=1.3.0 -I. -I../../src -I. -I../mumble_proto -I../../3rdparty/arc4random-src -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../release/.obj/murmur_grpcwrapper_protoc_plugin/main.o main.cpp
g++ -Wl,-z,relro -Wl,-z,now -Wl,-O1 -o ../../release/protoc-gen-murmur-grpcwrapper ../../release/.obj/murmur_grpcwrapper_protoc_plugin/main.o   -L/root/mumble/release -L/usr/lib/x86_64-linux-gnu/c++11 -lprotoc -lssl -lcrypto -ldl -lprotobuf -pthread -lz -lpthread
make[2]: Leaving directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
make[1]: Leaving directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
cd src/murmur/murmur_grpc/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur_grpc/murmur_grpc.pro 'CONFIG+=no-client grpc static' ) && make -f Makefile
make[1]: Entering directory '/root/mumble/src/murmur/murmur_grpc'
make -f Makefile.Release
make[2]: Entering directory '/root/mumble/src/murmur/murmur_grpc'
/usr/bin/protoc --cpp_out=. -I. -I.. ../MurmurRPC.proto
/usr/bin/protoc --grpc_out=. --plugin=protoc-gen-grpc= -I. -I.. ../MurmurRPC.proto
: program not found or is not executable
--grpc_out: protoc-gen-grpc: Plugin failed with status code 1.
make[2]: *** [Makefile.Release:208: MurmurRPC.grpc.pb.cc] Error 1
make[2]: Leaving directory '/root/mumble/src/murmur/murmur_grpc'
make[1]: Leaving directory '/root/mumble/src/murmur/murmur_grpc'
make[1]: *** [Makefile:42: release] Error 2
make: *** [Makefile:180: sub-src-murmur-murmur_grpc-make_first-ordered] Error 2
The command '/bin/sh -c make' returned a non-zero code: 2```

It looks like it can't find the `protoc-gen-grpc` plugin.

For reference, here's the Dockerfile I'm currently using:
```
FROM ubuntu:disco

# needed to install tzdata in disco
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get install -y \
    build-essential \
    pkg-config \
    qt5-default \
    libboost-dev \
    libasound2-dev \
    libssl-dev \
    libspeechd-dev \
    libzeroc-ice-dev \
    libpulse-dev \
    libcap-dev \
    libprotobuf-dev \
    protobuf-compiler \
    libprotoc-dev \
    libogg-dev \
    libavahi-compat-libdnssd-dev \
    libsndfile1-dev \
    libgrpc++-dev \
    libxi-dev \
    qtcreator

COPY . /root/mumble
WORKDIR /root/mumble

RUN qmake -recursive main.pro CONFIG+="no-client grpc static"
RUN make
```
I had to add `libprotoc-dev` to the dependency list (originally from https://wiki.mumble.info/wiki/BuildingLinux#1.3.x).

**Comment by gclawes** [comment]
Fixed by installing `protobuf-compiler-grpc`

**Comment by gclawes** [comment]
Latest issue is related to building a static binary:

```
...
usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/libIce.a(DynamicLibrary.o): in function `IceInternal::DynamicLibrary::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
(.text+0x195): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: cannot find -lcares
/usr/bin/ld: cannot find -lQt5Network
/usr/bin/ld: cannot find -lQt5Sql
/usr/bin/ld: cannot find -lQt5Xml
/usr/bin/ld: cannot find -lQt5DBus
/usr/bin/ld: cannot find -lQt5Core
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile.Release:359: ../../release/murmurd] Error 1
make[2]: Leaving directory '/root/mumble/src/murmur'
make[1]: *** [Makefile:42: release] Error 2
make[1]: Leaving directory '/root/mumble/src/murmur'
make: *** [Makefile:223: sub-src-murmur-make_first-ordered] Error 2
The command '/bin/sh -c make' returned a non-zero code: 2
```

**Comment by davidebeatrici** [comment]
Unfortunately Ubuntu doesn't provide a statically built version of Qt, however there's a Docker Hub image available which provides it on top of the latest version of Ubuntu: https://hub.docker.com/r/fffaraz/qt

**Comment by gclawes** [comment]
In that case, I think I'll give up on the static build for now and stick with upstream ubuntu, it's still a fairly small image.  I looked at alpine, and it looks like it doesn't have packages for ice or libspeechd (and probably others), so building with that would be tough.

**Comment by davidebeatrici** [comment]
Indeed.

**Comment by gclawes** [comment]
Interestingly, alpine builds with a bunch of features disabled: https://git.alpinelinux.org/aports/tree/community/mumble/APKBUILD

```
CONFIG+="no-speechd no-g15 no-update no-ice bundled-celt no-embed-qt-translations no-bundled-speex"
```

I'm guessing we want a more full-featured build for the official image.

**Comment by davidebeatrici** [comment]
The Alpine build is missing Ice, G15 and speechd.

The other options are there because it's a distribution package and thus it should use the system components (e.g. Speex), aside from `bundled-celt` which is not required because it's the default (opposite of `no-bundled-celt`).

**Comment by gclawes** [comment]
Gotcha.  I'm using this config to build: `CONFIG+="no-client grpc"`.  From what I can tell G15 doesn't even come into play for a server-only build.

Getting towards the finish line here.  It looks like `make install` isn't actually doing anything: 
```
root@43196fac1061:~/mumble# make install
cd src/mumble_proto/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/mumble_proto/mumble_proto.pro 'CONFIG+=no-client grpc' ) && make -f Makefile install
make[1]: Entering directory '/root/mumble/src/mumble_proto'
make -f Makefile.Release install
make[2]: Entering directory '/root/mumble/src/mumble_proto'
make[2]: Nothing to be done for 'install'.
make[2]: Leaving directory '/root/mumble/src/mumble_proto'
make[1]: Leaving directory '/root/mumble/src/mumble_proto'
cd src/murmur/murmur_ice/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur_ice/murmur_ice.pro 'CONFIG+=no-client grpc' ) && make -f Makefile install
make[1]: Entering directory '/root/mumble/src/murmur/murmur_ice'
make -f Makefile.Release install
make[2]: Entering directory '/root/mumble/src/murmur/murmur_ice'
make[2]: Nothing to be done for 'install'.
make[2]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
make[1]: Leaving directory '/root/mumble/src/murmur/murmur_ice'
cd src/murmur_grpcwrapper_protoc_plugin/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur_grpcwrapper_protoc_plugin/murmur_grpcwrapper_protoc_plugin.pro 'CONFIG+=no-client grpc' ) && make -f Makefile install
make[1]: Entering directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
make -f Makefile.Release install
make[2]: Entering directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
make[2]: Nothing to be done for 'install'.
make[2]: Leaving directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
make[1]: Leaving directory '/root/mumble/src/murmur_grpcwrapper_protoc_plugin'
cd src/murmur/murmur_grpc/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur_grpc/murmur_grpc.pro 'CONFIG+=no-client grpc' ) && make -f Makefile install
make[1]: Entering directory '/root/mumble/src/murmur/murmur_grpc'
make -f Makefile.Release install
make[2]: Entering directory '/root/mumble/src/murmur/murmur_grpc'
make[2]: Nothing to be done for 'install'.
make[2]: Leaving directory '/root/mumble/src/murmur/murmur_grpc'
make[1]: Leaving directory '/root/mumble/src/murmur/murmur_grpc'
cd src/murmur/ && ( test -e Makefile || /usr/lib/qt5/bin/qmake -o Makefile /root/mumble/src/murmur/murmur.pro 'CONFIG+=no-client grpc' ) && make -f Makefile install
make[1]: Entering directory '/root/mumble/src/murmur'
make -f Makefile.Release install
make[2]: Entering directory '/root/mumble/src/murmur'
make[2]: Nothing to be done for 'install'.
make[2]: Leaving directory '/root/mumble/src/murmur'
make[1]: Leaving directory '/root/mumble/src/murmur'
```

What from the `release/` directory actually needs to be installed?  I'll copy that over to a clean build stage.
```
root@43196fac1061:~/mumble# ls -l release/
total 13139
-rw-r--r-- 1 root root  761012 May  8 01:13 libmumble_proto.a
-rw-r--r-- 1 root root 3813224 May  8 01:13 libmurmur_grpc.a
-rw-r--r-- 1 root root 2184568 May  8 01:13 libmurmur_ice.a
-rwxr-xr-x 1 root root 6386336 May  8 01:15 murmurd
-rwxr-xr-x 1 root root   47120 May  8 01:13 protoc-gen-murmur-grpcwrapper
```

**Comment by davidebeatrici** [comment]
> From what I can tell G15 doesn't even come into play for a server-only build.

Correct, it's only for the client. It provides support for the LCD of certain Logitech devices, such as the G15 keyboard.

> What from the `release/` directory actually needs to be installed? I'll copy that over to a clean build stage.

Just `murmurd`.

**Comment by gclawes** [comment]
Got it running:
```
gclawes@enterprise ~/development/gclawes/mumble git:(docker*) $ docker run -it --rm murmur-test
<W>2019-05-08 03:44:52.450 SSL: OpenSSL version is 'OpenSSL 1.1.1b  26 Feb 2019'
<W>2019-05-08 03:44:52.451 Initializing settings from /home/murmur/.murmurd/murmur.ini (basepath /home/murmur/.murmurd)
<W>2019-05-08 03:44:52.482 MetaParams: TLS cipher preference is "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA:AES256-SHA:AES128-SHA"
<W>2019-05-08 03:44:52.484 ServerDB: Opened SQLite database /home/murmur/murmur.sqlite
<W>2019-05-08 03:44:52.484 ServerDB: Using SQLite's default rollback journal.
<W>2019-05-08 03:44:53.688 Performed initial PBKDF2 benchmark. Will use 16000 iterations as default
<W>2019-05-08 03:44:53.688 Generating new tables...
<W>2019-05-08 03:44:53.719 OSInfo: Failed to execute lsb_release
<W>2019-05-08 03:44:53.719 Murmur 1.3.0 (Compiled by User) running on X11: Linux 4.20.14-gentoo-r1: Booting servers
<W>2019-05-08 03:44:53.758 1 => Password for 'SuperUser' set to 'tjxbJYKHQyE2'
<W>2019-05-08 03:44:53.788 1 => Server listening on 0.0.0.0:64738
<W>2019-05-08 03:44:53.793 1 => Failed to set IPV6_RECVPKTINFO for 0.0.0.0:64738
<W>2019-05-08 03:44:53.838 1 => Generating new server certificate.
<W>2019-05-08 03:44:53.875 1 => Announcing server via bonjour
<W>2019-05-08 03:44:53.884 1 => Not registering server as public
```

What's the best way to configure the default directories (for murmur.ini, sqlite db, etc)?

**Comment by davidebeatrici** [comment]
You can specify the path to the configuration file with the `-ini` parameter.

**Comment by gclawes** [comment]
I'll set that in the default `CMD` for the container.

**Comment by davidebeatrici** [comment]
Of course!

**Comment by gclawes** [comment]
Cool.  My recommendation would be to use the `:latest` docker tag to mean "latest tagged release".  If you want a docker tag for the latest commit to master, `:snapshot` or `:edge` is usually the pattern I've seen for that.  We should probably manually push the first images then let DockerHub Builds handle any subsequent releases.

Once this is merged into master, this can be done with these commands:
```
$ git checkout 1.3.0-rc1
$ git checkout master -- Dockerfile
$ docker build -t mumble-voip/murmur:1.3.0-rc1 -t mumble-voip/murmur:latest .
$ docker push mumble-voip/murmur:1.3.0-rc1
$ docker push mumble-voip/murmur:latest
```

If you want to make me a collaborator on the DockerHub project (same username, gclawes), I can take care of this.

**Comment by davidebeatrici** [comment]
Thank you very much!

I created an organization and invited you, unfortunately I had to call it `mumblevoip` because `mumble` was already taken and only alphanumeric characters are allowed.

**Comment by gclawes** [comment]
Interesting.  It looks like the `mumble` organization links to `http://mumble.info/`.

Does it make sense to rename the image to `mumblevoip/murmur` to indicate that it's the server component only?

**Comment by davidebeatrici** [comment]
Definitely, maybe renaming it to `mumblevoip/murmur` would be even better.

**Comment by gclawes** [comment]
Can you also link the dockerhub repo (`mumblevoip/mumble-server`) to the `mumblevoip/mumble` GitHub account so I can configure auto-builds?  I think that makes more sense than linking to my personal account.

**Comment by davidebeatrici** [comment]
I created a dedicated account and connected it to the Docker Hub organization.

**Comment by gclawes** [comment]
I see `MumbleDockerBot` connected to the docker repo, but it doesn't seem to have access to the github repo to pull for builds:

![Capture](https://user-images.githubusercontent.com/172517/57497518-007f6f80-72a6-11e9-9b49-ffd01b5d9343.PNG)

**Comment by davidebeatrici** [comment]
It seems to be a Docker Hub issue, even with my GitHub account it doesn't show any repositories.

**Comment by gclawes** [comment]
I'll push the images for `mumblevoip/mumble:1.3.0-rc`/`mumblevoip/mumble:latest` manually for now

**Comment by gclawes** [comment]
I think this may be the procedure to grant DockerHub access to the `mumble-voip` GitHub organization: https://docs.docker.com/docker-hub/builds/link-source/#grant-access-to-a-github-organization

**Comment by davidebeatrici** [comment]
I granted access to the organization, but the issue persists.

**Comment by gclawes** [comment]
I'm not seeing the `mumble-voip` organization available in the builds menu, just `MumbleDockerBot`:

<img width="1116" alt="Screen Shot 2019-05-10 at 1 40 13 PM" src="https://user-images.githubusercontent.com/172517/57545706-344fa900-7329-11e9-90b7-c78353f969f0.png">

**Comment by davidebeatrici** [comment]
![](https://user-images.githubusercontent.com/5897523/57546002-1e0f1080-735c-11e9-8fde-ce59551dd618.png)

**Comment by krono86** [comment]
How can I find the official docker repo?
Thank you

**Comment by Krzmbrzl** [comment]
@krono86 the Dockerfile is here: https://github.com/mumble-voip/mumble/blob/master/Dockerfile

Is that what you meant?

**Comment by Avatat** [comment]
@krono86, we don't have an official Docker repository.
As @Krzmbrzl mentioned - you can use Dockerfile from Mumble's git repository, to build Docker image.

@Krzmbrzl, @davidebeatrici, @Kissaki, we should think about adding an official Murmur image to Docker Hub. I could try to do that.

**Comment by krono86** [comment]
An official Murmur image to Docker Hub would be great!

**Comment by Krzmbrzl** [comment]
> It seams that main.pro is not found.
What's happening?

I assume that a) the source tree isn't present or b) the qmake command is not run from the root of the Mumble-repo

**Comment by krono86** [comment]
Ok, it there was a problem with the rights.
Now I have an other error:

```
Makefile:453: recipe for target 'sub-src-murmur-release_ordered' failed
{standard input}: Assembler messages:
{standard input}:153802: Warning: end of file not at end of a line; newline inserted
{standard input}:154356: Error: invalid operands (*UND* and .gcc_except_table sections) for `-'
{standard input}:154372: Error: invalid operands (.text and *UND* sections) for `-'
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
make[2]: *** [../../release/.obj/murmur/MurmurGRPCImpl.o] Error 4
make[1]: *** [release] Error 2
make: *** [sub-src-murmur-release_ordered] Error 2
ERROR: Service 'mumble' failed to build: The command '/bin/sh -c make release' returned a non-zero code: 2
```

**Comment by Krzmbrzl** [comment]
> g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.

I guess you should do just that. Seems like you have encountered a bug in gcc

**Comment by toby63** [comment]
@Krzmbrzl Could you reopen this? In case someone wants to pick it up.

**Comment by Krzmbrzl** [comment]
@davidebeatrici what's the status of the Docker bot that is mentioned further up?

**Comment by Krzmbrzl** [comment]
@davidebeatrici what's the status of the Docker bot that is mentioned further up?

**Comment by theAkito** [comment]
There are a couple of images out there. I made one too and keep using it to this day. As it is the image I personally created, I obviously am of the opinion that my image is the best of all available. 😉 

https://github.com/theAkito/docker-murmur
https://hub.docker.com/r/akito13/murmur

P.S.:

To the maintainers:
I deeply apologise for not investigating the GRPC issue any further. I did not have any more time and dedication to continue. Perhaps some day I could test it again, but I can't make any promises soon.

**Comment by jdaviescoates** [comment]
I'm a bit confused by the status of this.  It there now an official maintained Docker for Mumble somewhere?

> @Krzmbrzl, @davidebeatrici, @Kissaki, we should think about adding an [official](https://docs.docker.com/docker-hub/official_images/) Murmur image to Docker Hub. I could try to do that.

Did you try to do that @Avatat ? :)

Thanks!

**Comment by Avatat** [comment]
@jdaviescoates, thank you for calling me out!

I have to refresh my Mumble/Murmur knowledge and if @Krzmbrzl and @davidebeatrici agree, I will sit with those Docker things (we have to polish our Dockerfile, do everything in compliance with best practices, and prepare a PR to Docker's GitHub.
During all that preparing stage, I can push images to my Docker Hub repository.

**Comment by jdaviescoates** [comment]
@Avatat you're most welcome! :)

I'd really love to have Mumble available as a package on Cloudron and there being an official Docker image would greatly help with that :)

**Comment by Krzmbrzl** [comment]
> I have to refresh my Mumble/Murmur knowledge and if Krzmbrzl and davidebeatrici agree, I will sit with those Docker things 

I don't see any argument against this. I just don't want to do it myself, since I have exactly 0 experience when it comes to Docker xD

**Comment by Avatat** [comment]
I've created a dedicated repository for this task: https://github.com/Avatat/mumble-docker-image/blob/main/Dockerfile
If we decide to ask for "official certification", we will have to move all that repository to the `mumble-voip` user on GitHub, and point to it in the official Docker process.

@Krzmbrzl, @davidebeatrici, three questions to you:
1. What is the best way, to find all dependencies required by a specific CMake target?
I'm asking in relation to my change, where I removed some APT packages (like `libasound2-dev`) which aren't needed by `cmake -Dclient=OFF ...`
2. I suppose that we don't need to use the Ninja build system inside the Docker image? It gives us nothing when used just one time per every image build?
3. What is the current build configuration (cmake parameters?) for Murmur?

**Comment by Krzmbrzl** [comment]
> What is the best way, to find all dependencies required by a specific CMake target?

I don't know if a way to do this. Afaik cmake does not know (or care) about dependencies.
Thus I would say you'll have to figure this out via trial&error :thinking: 

> I suppose that we don't need to use the Ninja build system inside the Docker image? It gives us nothing when used just one time per every image build?

To the best of my knowledge Ninja does not bring any (real) advantages over plain make for full builds. It only starts to show some advantages in incremental builds.
Personally I always use make anyways, so it definitely gets the job done :)

> What is the current build configuration (cmake parameters?) for Murmur?

Wdym? The cmake parameters to only build the server? That'd be `cmake -Dclient=OFF -Dserver=ON -Dice=ON -Dgrpc=OFF` (`grpc` could also be turned on, but I think for an "official" image this should be disabled as officially this is not stable and ready for use)

**Comment by theAkito** [comment]
@Avatat 

You can take [the image I created](https://github.com/theAkito/docker-murmur/blob/master/Dockerfile) as an example.

I tried to make it as minimal as it gets, but with GRPC enabled. However, the latter can of course be still disabled.

To make this build maintainable, up to date and generic enough for new releases, all you need to do is switch the commenting in these lines:

https://github.com/theAkito/docker-murmur/blob/master/Dockerfile#L37-L38

(One could even have several pipeline configurations, where images are built from the current release tag and from `master`, for example.)

Then you just need to let the CD pipeline programmatically define one of the needed ARGs:

https://github.com/theAkito/docker-murmur/blob/master/Dockerfile#L3-L4

That's it. The image may then be built by the pipeline on every commit/release or whatever. Besides the above adjustments, you only need to make the proper configurations for the CD pipeline.

**Comment by JuniorJPDJ** [comment]
I don't get why you guys use those big ubuntu or debian to make dockerfile. I'm using alpine based docker murmur setup for few years now.

**Comment by theAkito** [comment]
> I don't get why you guys use those big ubuntu or debian to make dockerfile. I'm using alpine based docker murmur setup for few years now.

Where? I have tried to make an alpine image and it simply did not work out. So I had to use `buster-slim` instead.

Could you point me to the repository showing an Alpine based `murmur` image, that can have features like `grpc` enabled?

**Comment by xM8WVqaG** [comment]
A dump of the more popular existing repos plus ones already mentioned in this thread. Most are based on Alpine. 

| Docker Hub | Repo | Docker Pulls | Repo Stars | Base Image |
| --- | --- | --- | --- | --- |
| <https://hub.docker.com/r/phlak/mumble> | <https://github.com/PHLAK/docker-mumble> | 1M+ | 59 stars | Alpine |
| <https://hub.docker.com/r/coppit/mumble-server> | <https://github.com/coppit/docker-mumble-server> | 1M+ | 18 stars | Phusion (Ubuntu) |
| <https://hub.docker.com/r/derenderkeks/murmur> | <https://www.github.com/DerEnderKeks/docker-murmur> | 1M+ | 3 stars | Alpine |
| <https://hub.docker.com/r/goofball222/murmur> | <https://github.com/goofball222/murmur> | 1M+ | 22 stars | Alpine |
| <https://hub.docker.com/r/mattikus/murmur> | <https://github.com/mattikus/docker-murmur> | 1M+ | 34 stars | busybox |
| <https://hub.docker.com/r/mcfio/mumble-server> | <https://github.com/mcfio/docker-mumble-server> | 100K+ | 0 stars | Alpine |
| <https://hub.docker.com/r/sudoforge/mumble-server> | <https://github.com/sudoforge/docker-mumble-server> | 100K+ | 36 stars | Alpine |
| <https://hub.docker.com/r/mumblevoip/mumble> | [_This repo?_](https://github.com/mumble-voip/mumble/blob/master/Dockerfile) | 433 | 3.8k stars(?) | Ubuntu |
| <https://hub.docker.com/r/akito13/murmur> | <https://github.com/theAkito/docker-murmur> | 53 | 0 stars | Debian |

`grpc` isn't a blocker based on [Krzmbrzl's above comment](https://github.com/mumble-voip/mumble/issues/3688#issuecomment-810896814):

> `grpc` could also be turned on, but I think for an "official" image this should be disabled as officially this is not stable and ready for use

**Comment by JuniorJPDJ** [comment]
> Could you point me to the repository showing an Alpine based `murmur` image, that can have features like `grpc` enabled?
I never used grpc so I don't know that tbh. I'll post mine Dockerfile then.

It packages mumble as .apk file for alpine and installs it.

```
FROM    alpine:edge

RUN     apk -U add alpine-sdk sudo

RUN     mkdir -p /var/cache/distfiles && \
        adduser -D packager && \
        addgroup packager abuild && \
        chgrp abuild /var/cache/distfiles && \
        chmod g+w /var/cache/distfiles && \
        echo "packager    ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers && \
        mkdir -p /target/home/x86_64 /home/packager/.abuild && \
        chown -R packager:packager /home/packager/.abuild/ /target/home/x86_64

USER    packager

RUN     abuild-keygen -a -i
ADD     mumble-apkbuild /home/packager/
WORKDIR /home/packager

RUN     abuild -c -r -P /target


FROM    alpine:latest

COPY    --from=0 /target/home/x86_64/murmur-1.3.0-r7.apk .
RUN     set -ex; apk add --no-cache --allow-untrusted ./murmur-1.3.0-r7.apk qt5-qtbase-postgresql 

RUN     adduser -DHs /sbin/nologin mumble

EXPOSE  64738 64738/udp

USER    mumble

VOLUME  /etc/mumble

CMD     ["/usr/bin/murmurd", "-fg", "-ini", "/etc/mumble/config.ini"]
```

`mumble-apkbuild` is directory near Dockerfile containing APKBUILD for mumble (slightly modified version from official alpine aports)
I'm using 1.3.0 mumble server bcs of #4292

**Comment by Avatat** [comment]
Wow! Thank you for **huge** feedback :)

I didn't mention it, but this is a very early stage of work on the official Docker image.
I plan to prepare a CD pipeline to build every new release without the need of doing much manual work. I also plan to create Alpine-based Docker image.

**Comment by theAkito** [comment]
@xM8WVqaG 

Half of them use Debian based distributions and mine is the only one doing it right: compiling the server on build, not just downloading it and placing it into the image. That's what the magic is about. So the existing images don't say anything about how to change my image to `alpine`.



> I never used grpc so I don't know that tbh. I'll post mine Dockerfile then.

So, I can understand how you figured out how to use `alpine` as a base.

If you want to use all the features or at least have the possbility to use them, then you need to use a Debian based distribution, or you probably have to build the necessary dependencies from scratch to make it work on `alpine`. That is the point I got to, when trying to build an `alpine` image for `murmur`, 10 months ago.



> I plan to prepare a CD pipeline to build every new release without the need of doing much manual work. I also plan to create Alpine-based Docker image.

I'll just throw my 2 cents in again, if you don't mind:

* Make several configurations; for example, make one that builds frequently, I don't know if "on every commit" is too often, because of some limitations on the monthly allowance, however if I were in your shoes, I would still make at least 2 configurations, one of them building on a Github release.
* It's not uncommon to create `alpine` based images **and** `Debian` based images. Perhaps, if you arrive at the point I arrived at 10 months ago, you won't find a suitable way to build the image on `alpine` properly, leading you to having two options, where the `alpine` based image is for those who care more about how slim the result is, instead of how many features it has, while the `Debian` based image can be the one containing optional features. Something like that. (Or perhaps the release management can offer builds with features enabled, so the `alpine` based images can download those binaries on build.)

Of course, there is probably some way to do all that with `alpine` manually (without downloading a pre-compiled `murmur` binary), however that probably presumes quite some manual work, I suppose, so I don't know if that's worth it just to get it running in `alpine`. When I arrived at that point, I didn't see it as worth it to make the image so complicated and extremely custom just for the sake of running it on `alpine`. I mean, this distribution is a great container distribution, but not that great, that I would recommend anyone spend so much extra time just to get their containers running on that, especially when we are talking about `murmur`, where a single user probably would only use a single one on their machine, anyway.

**Comment by jdaviescoates** [comment]
>  I also plan to create Alpine-based Docker image.

I don't know enough about Docker to know, but would a Alpine-based Docker image also run on an Ubuntu VPS? (I'm guessing not).

Because one of the reasons I bumped this is I'd like there to be an official Docker image for Mumble so that it can more easily be packaged and maintained for Cloudron (which currently only runs on Ubuntu).

**Comment by JuniorJPDJ** [comment]
> I don't know enough about Docker to know, but would a Alpine-based Docker image also run on an Ubuntu VPS? (I'm guessing not).

Yes, It will. Docker images are packaging whole OS, that's why using slimmer images are good idea. You don't need to download another ubuntu just for mumble ;D

**Comment by Avatat** [comment]
Sorry for the delay.
I remember about Docker things and I will continue it through the next week.

**Comment by JuniorJPDJ** [comment]
@theAkito what deps are missing in alpine? I possibly could port them and add to alpine official edge repo.

**Comment by Avatat** [comment]
@JuniorJPDJ, there is no `libzeroc-ice`, and I have issues with gRPC building:
```
CMake Error at /usr/lib/cmake/grpc/gRPCTargets.cmake:685 (message):
  The imported target "gRPC::grpc_cpp_plugin" references the file

     "/usr/bin/grpc_cpp_plugin"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/lib/cmake/grpc/gRPCTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/lib/cmake/grpc/gRPCConfig.cmake:21 (include)
  cmake/pkg-utils.cmake:111 (find_package)
  src/murmur/CMakeLists.txt:197 (find_pkg)
```

**Comment by theAkito** [comment]
@JuniorJPDJ 

That'd be great!

I can't remember physically what exactly was missing, but as @Avatat already said, it was related to gRPC.

If you want to know the whole story, here are the references for understanding the issue:

https://github.com/mumble-voip/mumble/issues/4201

As far as I can see, I documented the process in a sufficient degree of detail, so I hope it helps. 😃

**Comment by JuniorJPDJ** [comment]
> I have issues with gRPC building
You installed `grpc-dev` i `grpc` ?

**Comment by Avatat** [comment]
I installed `grpc-dev` but not `grpc`. After I had installed `grpc`, I get the new error:
```
 ---> Running in 203f33abb8b7
...
-- Found ZLIB: /lib/libz.so (found version "1.2.11") 
-- Could NOT find c-ares (missing: c-ares_DIR)
-- Could NOT find c-ares (missing: c-ares_INCLUDE_DIR c-ares_LIBRARY) 
-- gRPC found | Version: 1.34.1
-- Configuring done
CMake Error at src/murmur/CMakeLists.txt:55 (add_executable):
  Target "murmur" links to target "c-ares::cares" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Do I guess properly, that c-ares is required by gRPC?

Current Dockerfile:
```
FROM alpine:3.13 as builder

RUN apk add boost-dev \
	build-base \
	c-ares \
	cmake \
	git \
	grpc \
	grpc-dev \
	libcap-dev \
	protobuf-dev \
	qt5-qtbase-dev \
	qt5-qttools-dev

RUN git clone -b master https://github.com/mumble-voip/mumble.git /root/mumble/

WORKDIR /root/mumble/build
RUN cmake -Dclient=OFF -DCMAKE_BUILD_TYPE=Release -Dgrpc=ON -Dzeroconf=OFF -Dice=OFF ..
RUN make -j $(nproc)
```

**Comment by Kissaki** [comment]
Can we move this into an open ticket? For what's still open and being worked on?

Given that the grpc implementation is supposedly broken, and for that reason also not enabled by default in Mumble builds, I assume we probably do not want support it in prebuilt Docker images.

https://github.com/mumble-voip/mumble/labels/gRPC

**Comment by JuniorJPDJ** [comment]
@Avatat if you are building something on Alpine you always need package without suffix and one with `-dev` as it contains headers and symbols. Now you are probably missing `c-ares-dev`.

**Comment by Avatat** [comment]
I'm sorry...
I thought that `c-ares-dev` will be installed, because it is a sub-package of `c-ares`... I'm stupid...

Now everything builds fine! Thank you @JuniorJPDJ! I will test Alpine image and will push it to my repo.
But, there is still an issue with lack of `zeroc-ice`. Do you know how to port it? Can you do this?

@Kissaki, can we open this issue? Can I ask for assigning it to me?

**Comment by JuniorJPDJ** [comment]
>  But, there is still an issue with lack of zeroc-ice. Do you know how to port it? Can you do this?

I'll try to port it later today ;D

**Comment by theAkito** [comment]
@Avatat 
@JuniorJPDJ 

Great job, guys! 👏🏻

**Comment by JuniorJPDJ** [comment]
It would be cool :D 
First stage build on some distro and then static build!
I'm worrying about db support tho, for pgsql support I'm using I had to install additional Qt support packages on alpine, I'm not sure how hard and if possible it would be on static build without distro.

**Comment by JuniorJPDJ** [comment]
We could provide 3 images: eg. master, master-alpine and master-ubuntu for different people and that would solve most of issues for different people wanting to base on those images and add their own changes on top.
The same of course would go for tags/releases, snapshots etc.

**Comment by Avatat** [comment]
That is the plan :)

**Comment by Krzmbrzl** [comment]
@Avatat https://github.com/mumble-voip/mumble/blob/master/docs/dev/build-instructions/build_static.md

**Comment by the4anoni** [comment]
Hi
Will ARM/ARM64 be supported?

**Comment by pataquets** [comment]
On recent Docker versions you can build directly from source on Github:
```
docker build -t mumble-server github.com/mumble-voip/mumble
```
I'm not sure if/how you can checkout specific tags/branches.
@Avatat @davidebeatrici @Krzmbrzl : If you like this method documented, let me know where it would be fit, so I can send a PR.

**Comment by JuniorJPDJ** [comment]
> I'm not sure if/how you can checkout specific tags/branches.

Using hash character after address:
docs.docker.com/engine/reference/commandline/build/#git-repositories

**Comment by Krzmbrzl** [comment]
@pataquets I think at this point the best position for this would be directly in the README at the end of [this section](https://github.com/mumble-voip/mumble#running-murmur) :thinking:

**Comment by pataquets** [comment]
Done. Sent PR #5258.
@Krzmbrzl @JuniorJPDJ : BTW, it would be great to merge recent Dockerfile build improvements into `1.4.x` branch (which currently breaks), thus enabling easier testing of current devel branch.

**Comment by Krzmbrzl** [comment]
#5259

**Comment by jdaviescoates** [comment]
I see there has been a bunch of activity and PRs here, and that this is now closed...so I guess that means there is now an official Docker image? I thought perhaps I'd see something over at https://hub.docker.com/u/mumblevoip but nothing new there.

**Comment by Krzmbrzl** [comment]
The Docker image is now managed in https://github.com/mumble-voip/mumble-docker

**Comment by nRaecheR** [comment]
https://hub.docker.com/r/mumblevoip/mumble-serv gives only a 404, so there are no official images. Yes, it's possible to build it from sources but it would be better to provide one. The link is from the README.md from GitHub.

**Comment by Krzmbrzl** [comment]
> gives only a 404, so there are no official images.

Not _yet_. We are in the process of setting up GitHub Actions for automatic pushes to Dockerhub. The repo i still WIP but it should be finished in the upcoming days, I think

**Comment by fuuman** [comment]
Are there any updates on providing the official mumble image on the docker hub?

**Comment by the4anoni** [comment]
> > gives only a 404, so there are no official images.
> 
> Not _yet_. We are in the process of setting up GitHub Actions for automatic pushes to Dockerhub. The repo i still WIP but it should be finished in the upcoming days, I think

Any news on this?

**Comment by Krzmbrzl** [comment]
https://hub.docker.com/r/mumblevoip/mumble-server

But also consider https://github.com/mumble-voip/mumble/issues/5613

**Comment by danger89** [comment]
Thx!
